lib/handler/headers_util.c

```diff
@@ -1,29 +1,109 @@
 #include "h2o.h"
 
-static h2o_header_t *find_header(h2o_headers_t *headers, h2o_headers_command_t *cmd)
+static h2o_header_t *find_header(h2o_headers_t *headers, h2o_iovec_t *name)
 {
     ssize_t index;
 
-    if (h2o_iovec_is_token(cmd->name)) {
-        index = h2o_find_header(headers, (void *)cmd->name, -1);
+    if (h2o_iovec_is_token(name)) {
+        index = h2o_find_header(headers, (void *)name, -1);
     } else {
-        index = h2o_find_header_by_str(headers, cmd->name->base, cmd->name->len, -1);
+        index = h2o_find_header_by_str(headers, name->base, name->len, -1);
     }
     if (index == -1)
         return NULL;
     return headers->entries + index;
 }
 
+static int is_in_list(const char *base, size_t len, h2o_headers_command_t *cmd)
+{
+    size_t i;
+    h2o_iovec_t name = h2o_iovec_init(base, len);
+    for (i = 0; i != cmd->num_args; ++i) {
+        if (h2o_iovec_is_token(cmd->args[i].name)) {
+            if (cmd->args[i].name->base == name.base) {
+                return 1;
+            }
+        } else {
+            if (h2o_memis(cmd->args[i].name->base, cmd->args[i].name->len, name.base, name.len))
+                return 1;
+        }
+    }
+    return 0;
+}
+
+static void filter_cookie(h2o_mem_pool_t *pool, char **base, size_t *len, h2o_headers_command_t *cmd)
+{
+    h2o_iovec_t iter = h2o_iovec_init(*base, *len), token_value;
+    const char *token;
+    size_t token_len;
+    char dst[*len * 2];
+    size_t dst_len = 0;
+
+    do {
+        if ((token = h2o_next_token(&iter, ';', ';', &token_len, &token_value)) == NULL)
+            break;
+        int found = is_in_list(token, token_len, cmd);
+        if ((cmd->cmd == H2O_HEADERS_CMD_COOKIE_UNSETUNLESS && found) || (cmd->cmd == H2O_HEADERS_CMD_COOKIE_UNSET && !found)) {
+            if (dst_len != 0) {
+                memcpy(dst + dst_len, H2O_STRLIT("; "));
+                dst_len += 2;
+            }
+            memcpy(dst + dst_len, token, token_len);
+            dst_len += token_len;
+            if (token_value.len > 0) {
+                memcpy(dst + dst_len, H2O_STRLIT("="));
+                dst_len++;
+                memcpy(dst + dst_len, token_value.base, token_value.len);
+                dst_len += token_value.len;
+            }
+        }
+    } while (1);
+    if (dst_len > *len) {
+        if (pool == NULL) {
+            free(*base);
+            *base = malloc(dst_len);
+        } else {
+            *base = h2o_mem_alloc_pool(pool, *dst, dst_len);
+        }
+    }
+    memcpy(*base, dst, dst_len);
```

Comment:
I do not think there is a need to use a temporary variable (`i.e. `char dst[*len * 2]`) then copy the result to a different vector.

As the `h2o_rewrite_headers` function is assumed to be called as part of request processing, there is a guarantee that `pool` would be non-NULL.

Therefore, we can do a `h2o_mem_alloc(pool, len * 2)` at the beginning of the function, build string on that buffer, then return that buffer as the new header field value.

Reply:
Fair enough. It was handy to support the NULL case because i tested the code in a fuzzer. I'll remove it.

Reply:
Fixed in 3fd79f582